specification/sections/05-events.md

```diff
@@ -0,0 +1,102 @@
+
+## Overview
+
+`Events` allow consumers (_application integrator_, _application author_, _integration author_) to react to state changes in the provider or underlying flag management system, such as flag definition changes, provider readiness, or error conditions. A provider may emit events or run a callback indicating that it received a certain event, optionally providing data associated with that event. Handlers registered on the client are then invoked with this data.
+
+### 5.1. Provider events
+
+#### Requirement 5.1.1
+
+> The `provider` **MAY** define a mechanism for signaling the occurrence of one of a set of events, including `PROVIDER_READY`, `PROVIDER_ERROR`, `PROVIDER_CONFIGURATION_CHANGED` and `PROVIDER_STALE`, with an `provider event details` payload. 
+
+If available, native event-emitter or observable/observer language constructs can be used.
+
+see: [provider event types](../types.md#provider-events), [`event details`](../types.md#provider-event-details).
+
+#### Requirement 5.1.2
+
+> When the `provider` signals the occurrence of a particular `event`, the associated `client` and `API` event handlers` **MUST** run.
+
+see: [provider event types](./../types.md#provider-events) and [event handlers](#52-event-handlers).
+
+#### Requirement 5.1.3
+
+> `PROVIDER_ERROR` events **SHOULD** populate the `provider event details`'s `error message` field.
+
+The error message field should contain an informative message as to the nature of the error.
+
+See [event metadata](../types.md#error-event-details)
+
+### 5.2. Event handlers
+
+#### Requirement 5.2.1
+
+> The `client` **MUST** provide an `addHandler` function for associating callbacks with `provider events`, which accepts an `event type` and a `event handler function`.
+
+```java
+  // run the myClientOnReadyHandler function when the PROVIDER_READY event is fired
+  client.addHandler(ProviderEvents.Ready, MyClass::myClientOnReadyHandler);
+```
+
+see: [provider events](#51-provider-events)
+
+#### Requirement 5.2.2
+
+> The `API` **MUST** provide an `addHandler` function for associating callbacks with `provider events`, which accepts an `event type` and a `event handler function`.
+
+```java
+  // run the myGlobalErrorHandler function when the PROVIDER_READY event is fired
+  OpenFeature.addHandler(ProviderEvents.Error, MyClass::myGlobalErrorHandler);
+```
+
+see: [provider events](#51-provider-events), [`provider event types`](../types.md#provider-events)
+
+#### Requirement 5.2.3
+
+> The `event details` **MUST** contain the `client name` associated with the event.
+
+The `client name` indicates the client/provider with which the event is associated.
+This is particularly relevant for `event handler functions` which are attached to the `API`, not a particular client.
+
+#### Requirement 5.2.4
+
+> The `event handler` function **MUST** accept a `event details` or `error event metadata` parameter.
```

Comment:
The usage of `or` is confusing here. Taking Java as an example, in minimal form, the handler can be a functional interface.

```
public interface EventHandler {
    void handle(EventDetails details);
}
```

Here, I expect `EventDetails` to be common for any event type. Otherwise, I believe language-specific restrictions will make it difficult to expose a single `addHandler(EventHandler)` method as the method signature will change.

Reply:
Hey @Kavindu-Dodan , you're right. The thing I was trying to do here was to note that events could contain the `error event details` or the `event details`... but I've consolidated both of these into a single type which has a message. The message could be useful in both error and non-error scenarios, so I think the distinction isn't worth the complexity.

Please see delta [here](https://github.com/open-feature/spec/pull/182/commits/e09a20ae9906a05e1aed8ca265ca7ceb4e3d7e21).

Reply:
Thanks I am having another look into this :)